include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td

```diff
@@ -360,18 +383,24 @@ def WireOp : FIRRTLOp<"wire", []> {
     ```
     }];
 
-  let arguments = (ins StrAttr:$name, AnnotationArrayAttr:$annotations);
+  let arguments = (
+    ins StrAttr:$name, AnnotationArrayAttr:$annotations,
+        OptionalAttr<SymbolNameAttr>:$inner_sym);
```

Comment:
Does it make sense to unify the `name` and `inner_sym` ? The `name` also has to be unique within the module. Maybe we can always add the `name` as the `inner_sym` ? 
It might make the builder simpler, and also lowering simpler to just use the `inner_sym` ?
This comment is true for all the operations

Reply:
okay, that makes sense. Then the question is when do we add the `inner_sym` in FIRRTL, the parser never adds the symbol, so which pass introduces the `inner_sym`?
 Also is there a case where an op has a name but no `inner_sym`, and hence no symbol is generated for the lowered op. But previously, just the existence of a name would have resulted in an `inner_sym` after lowering. Looking at the `LowerToHw` seems like this case doesn't occur.

Reply:
The `inner_sym` is optional at the moment. The intent is that if a pass needs a symbol for one of these operations, it can quickly conjure one up if there isn't one already, and then use that. This is going to be the case for a lot of the OMIR/metadata emission and Grand Central business, which will want to refer to instances and the like. Those passes will add `inner_sym`s liberally to get a handle on the ops.

Since the `inner_sym` is optional (and just for IR-internal use), it's very common for ops to have a name but no symbol. What happens in `LowerToHW` is the following:

- If the FIRRTL op had an `inner_sym`, the HW op will inherit that
- If the FIRRTL op had no symbol but is marked "don't touch", the HW op will get a new `inner_sym`
- Otherwise the HW op gets no `inner_sym`

The "don't touch" part is how `LowerToHW` handled things prior to this patch, where it would add a symbol just to prevent further modification/removal of the HW op. This PR adds an additional way how the resulting HW op can get a symbol.

I don't think the existence of a name resulted in an `inner_sym` in `LowerToHW` though. If there's no "don't touch" attribute present, a `firrtl.wire` would just lower to a `sv.wire` without any symbol. Only if a "don't touch" was present would the lowering conjure up a symbol for the resulting wire: https://github.com/llvm/circt/blob/6608f5b9b351e68ab53c9a00790e4225a740a547/lib/Conversion/FIRRTLToHW/LowerToHW.cpp#L1836-L1866

Reply:
annotation scattering will need to add symbols to things it resolves paths too.  Once nla take inner_name_refs instead of names, that will be the only choice anyway.